### PR TITLE
Unify validation and execution: Deposit

### PR DESCRIPTION
**Pull request by itamarreif**
This PR refactors verification for the deposit function such that only transaction-level validation is done on the entire batch, and per-deposit validation is moved to a separate function.

Related to issue #194.

### CODE REVIEW

File: x/dex/types/message_deposit.go [hunk 1]
```diff
@@ -44,9 +44,27 @@ func (msg *MsgDeposit) GetSignBytes() []byte {
 }
 
 func (msg *MsgDeposit) ValidateBasic() error {
+	// Converts input address (string) to sdk.AccAddress
```
Comment (teddyknox) [inline x/dex/types/message_deposit.go:47]: Self-explanatory, can remove

File: x/dex/types/message_deposit.go [hunk 2]
```diff
@@ -44,9 +44,27 @@ func (msg *MsgDeposit) GetSignBytes() []byte {
...
 	_, err := sdk.AccAddressFromBech32(msg.Creator)
+	// Error checking for the calling address
```
Comment (teddyknox) [inline x/dex/types/message_deposit.go:48]: Self-explanatory, can remove

File: x/dex/keeper/core_helper.go [hunk 3]
```diff
@@ -521,3 +526,61 @@ func (k Keeper) BurnShares(ctx sdk.Context, addr sdk.AccAddress, amount sdk.Int,
...
+	ctx sdk.Context,
+	callerAddr sdk.AccAddress,
+	feeTiers []types.FeeTier,
+	feeIndex uint64,
+	pair *types.TradingPair,
+	tickIndex int64,
+	amount0 sdk.Int,
+	amount1 sdk.Int,
+	balance0 sdk.Int,
+	balance1 sdk.Int,
```
Comment (teddyknox) [inline x/dex/keeper/core_helper.go:544]: This signature is too big. Why are we separating this into a separate function?

File: x/dex/keeper/core.go [hunk 4]
```diff
@@ -118,6 +109,8 @@ func (k Keeper) DepositCore(
 		amounts1Deposited[i] = inAmount1
 		totalAmountReserve0 = totalAmountReserve0.Add(inAmount0)
 		totalAmountReserve1 = totalAmountReserve1.Add(inAmount1)
+		balance0 = balance0.Sub(inAmount0)
+		balance1 = balance1.Sub(inAmount1)
```
Comment (teddyknox) [inline x/dex/keeper/core.go:113]: what's going on here?

File: x/dex/keeper/core.go [hunk 5]
```diff
@@ -42,27 +42,18 @@ func (k Keeper) DepositCore(
...
-
-		// TODO: Allow user to deposit "behind enemy lines"
-		if amount0.GT(sdk.ZeroInt()) && curTick0to1 <= lowerTickIndex {
-			return nil, nil, types.ErrDepositBehindPairLiquidity
-		}
-
-		// TODO: Allow user to deposit "behind enemy lines"
-		if amount1.GT(sdk.ZeroInt()) && upperTickIndex <= curTick1to0 {
-			return nil, nil, types.ErrDepositBehindPairLiquidity
+		fee, lowerTickIndex, upperTickIndex, err := k.ValidateDeposit(ctx, callerAddr, feeTiers, feeIndex, &pair, tickIndex, amount0, amount1, balance0, balance1)
```
Comment (teddyknox) [inline x/dex/keeper/core.go:54]: Let's chop down lines like these that go off the page

Comment (teddyknox) [reply x/dex/keeper/core.go:54]: Potentially not have a separate call here

File: x/dex/keeper/core_helper.go [hunk 6]
```diff
@@ -521,3 +526,61 @@ func (k Keeper) BurnShares(ctx sdk.Context, addr sdk.AccAddress, amount sdk.Int,
...
+) (int64, int64, int64, error) {
+	// check that feeIndex is a valid index of the fee tier
```
Comment (teddyknox) [inline x/dex/keeper/core_helper.go:546]: Do we have a test case for each of these checks?

**Review by teddyknox — COMMENTED** [review]
Do we have a test case for each che

File: x/dex/keeper/core_helper.go (hunk 3 above)
Comment (itamarreif) [reply x/dex/keeper/core_helper.go:544]: Seemed cleaner to me to have this in a separate function. Can move it back if you want.

File: x/dex/keeper/core.go (hunk 4 above)
Comment (itamarreif) [reply x/dex/keeper/core.go:113]: This is for checking the remaining balance after a deposit (midway through the batch).

i.e. balance0 after 1st deposit = balance0 - first inAmount0
balance0 after 2nd deposit = balance0 - (first inAmount0) - (second inAmount0)
etc.

File: x/dex/keeper/core_helper.go (hunk 6 above)
Comment (itamarreif) [reply x/dex/keeper/core_helper.go]: Added

File: x/dex/keeper/core.go (hunk 4 above)
Comment (teddyknox) [reply x/dex/keeper/core.go:147]: Doesn't the bank account for whether a user has enough tokens for the transaction?

File: x/dex/keeper/verification.go [hunk 7]
```diff
@@ -9,77 +9,20 @@ import (
 )
 
 func (k Keeper) DepositVerification(goCtx context.Context, msg types.MsgDeposit) (string, string, sdk.AccAddress, []sdk.Int, []sdk.Int, error) {
```
Comment (teddyknox) [inline x/dex/keeper/verification.go:11]: Are we not getting rid of this?

File: x/dex/keeper/core.go (hunk 4 above)
Comment (itamarreif) [reply x/dex/keeper/core.go:147]: the transfer is only made at the end of the transaction so bank balances arent updated until then.

File: x/dex/keeper/verification.go (hunk 7 above)
Comment (itamarreif) [reply x/dex/keeper/verification.go:11]: moved sorting and msg validation call to msg server

File: x/dex/keeper/core.go (hunk 4 above)
Comment (teddyknox) [reply x/dex/keeper/core.go:147]: Yes, my point: do we need validation here?

File: x/dex/keeper/core.go [hunk 8]
```diff
@@ -43,28 +43,43 @@ func (k Keeper) DepositCore(
...
 
-	FeeTier := k.GetAllFeeTier(ctx)
+	feeTiers := k.GetAllFeeTier(ctx)
 
 	for i, amount0 := range amounts0 {
 		amount1 := amounts1[i]
 		tickIndex := msg.TickIndexes[i]
 		feeIndex := msg.FeeIndexes[i]
-		fee := FeeTier[feeIndex].Fee
-		curTick0to1 := pair.CurrentTick0To1
```
Comment (teddyknox) [inline x/dex/keeper/core.go:53]: why are we removing these?

**Review by teddyknox — APPROVED** [review]

**Pull request merged**